Feature/enhanced markdown rendering support

kocheick opened the pull request:
## Description
Enhanced markdown support with comprehensive footnotes and improved task lists functionality.

## Changes Made
- **Footnotes Support**: Added complete footnotes rendering in KotlinRenderer with proper ID-based linking
- **Enhanced Task Lists**: Improved task list detection and rendering in all text contexts, including table cells
- **Feature Configuration**: Added footnotes feature toggle in MarkdownFeatures
- **Playground Examples**: Updated test examples demonstrating new features

## Problem Solved
This PR addresses inconsistencies in markdown rendering, specifically:
- Missing footnotes support (GitHub Flavored Markdown feature)
- Inconsistent task list rendering in complex structures like tables
- Need for better markdown feature extensibility

## Testing
- [ ] Tested footnotes rendering with references and definitions
- [ ] Verified task list rendering with and without Silk
- [ ] Confirmed backwards compatibility
- [ ] Tested in playground environment

## Type of Change
- [x] New feature (non-breaking change which adds functionality)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

Fixes: Enhanced markdown rendering capabilities

bitspittle committed:
Update Kobweb CLI version to v0.9.20

bitspittle committed:
Update CLI version to 0.9.21

kocheick committed:
feat: Enhanced markdown support with footnotes and improved task lists

- Add comprehensive footnotes support in KotlinRenderer
  - Footnote references rendered as superscript links
  - Footnote definitions rendered as block elements
  - Proper ID-based linking between references and definitions

- Enhance task list support in MarkdownHandlers
  - Automatic detection of task list patterns in all text contexts
  - Support for - [x] (checked) and - [ ] (unchecked) syntax
  - FontAwesome icons when Silk is available, HTML checkboxes fallback
  - Works in table cells and other complex structures

- Add footnotes feature configuration in MarkdownFeatures
  - Configurable footnotes support (enabled by default)
  - Integration with CommonMark FootnotesExtension

- Update playground with test examples for new features

Resolves issues with markdown rendering inconsistencies and adds
support for GitHub Flavored Markdown footnotes feature.

kocheick committed:
feat: Enhanced markdown support with footnotes and improved task lists

- Add comprehensive footnotes support in KotlinRenderer
  - Footnote references rendered as superscript links
  - Footnote definitions rendered as block elements
  - Proper ID-based linking between references and definitions

- Enhance task list support in MarkdownHandlers
  - Automatic detection of task list patterns in all text contexts
  - Support for - [x] (checked) and - [ ] (unchecked) syntax
  - FontAwesome icons when Silk is available, HTML checkboxes fallback
  - Works in table cells and other complex structures

- Add footnotes feature configuration in MarkdownFeatures
  - Configurable footnotes support (enabled by default)
  - Integration with CommonMark FootnotesExtension

- Update playground with test examples for new features

Resolves issues with markdown rendering inconsistencies and adds
support for GitHub Flavored Markdown footnotes feature.

kocheick committed:
add support for github flavored markdown footnotes and enhance task list renderingtency and features.

kocheick committed:
add support for github flavored markdown footnotes and enhance task list renderingtency and features.

kocheick committed:
Merge branch 'main' into feature/enhanced-markdown-support

bitspittle commented on playground/site/src/jsMain/resources/markdown/documents/unified_test.md:
```diff
@@ -0,0 +1,102 @@
+posted: true
+routeOverride: /tests/
+
+#layout: .components.layouts.EnhancedMarkdownLayout
```
Nit: Remove stale layout reference?

kocheick replied:
good point!

bitspittle commented on playground/site/src/jsMain/resources/markdown/documents/unified_test.md:
```diff
@@ -0,0 +1,102 @@
+description: Testing unified markdown processing with remark-gfm and rehype-raw plugins
+tags: [ 'kobweb', 'commonmark', 'gfm', 'test', 'md', 'kotlin' ]
+posted: true
+routeOverride: /tests/
```
Maybe put this under /markdown/tests ?

kocheick replied:
sure thing

bitspittle commented on tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/KotlinRenderer.kt:
```diff
@@ -411,7 +416,9 @@ class KotlinRenderer internal constructor(
         }
 
         override fun visit(listItem: ListItem) {
+
```
Nit: Remove these extra spaces

bitspittle commented on tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/MarkdownFeatures.kt:
```diff
@@ -100,13 +102,35 @@ abstract class MarkdownFeatures {
+     * @see <a href="https://github.com/commonmark/commonmark-java#strikethrough">Strikethrough</a>
+     */
+    @get:Input
+    abstract val strikethrough: Property<Boolean>
```
Nit: Let's keep these alphabetized for consistency as well.

kocheick replied:
you're referring to strikethrough coming before tables stuff as in ? :

```
    ...
    abstract val strikethrough: Property<Boolean>
    abstract val table...
```

bitspittle replied:
Yes, and same with the code below:

```kotlin
if (strikethrough.get()) {
   extensions.add(StrikethroughExtension.create())
}
if (footnotes.get()) {
   extensions.add(FootnotesExtension.create())
}
```

kocheick replied:
understood💪

bitspittle commented on tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/MarkdownFeatures.kt:
```diff
@@ -100,13 +102,35 @@ abstract class MarkdownFeatures {
+     *
+     * Defaults to `true`.
+     *
+     * Note: Inline footnotes via "^[inline]" are not enabled by default.
```
This note makes it sound like there's a way to enable inline footnotes, but I'm guessing we don't support that right now at all?

If so, I might write this as "Note: Inline footnotes via "^[inline]" are not supported at this time."

bitspittle commented on tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/handlers/MarkdownHandlers.kt:
```diff
@@ -258,7 +293,53 @@ abstract class MarkdownHandlers @Inject constructor(project: Project) {
+            // Renders as:
+            // - With Silk: FontAwesome checkbox icons with proper spacing
+            // - Without Silk: HTML input checkboxes (disabled) with proper checked state
+            val taskListPattern = Regex("""^-\s*\[([x\s])\](.*)$""")
```
I'm not 100% sure yet but using a Regex here feels wrong.

The way most markdown features work is the parser breaks up the text and exposes rich-typed children nodes to us at this point.

Are you sure there's no other way?

kocheick replied:
fair enough, it gets the job done but definitely agree it feels it feels out of bound somehow

bitspittle commented on tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/KotlinRenderer.kt:
```diff
@@ -483,6 +473,58 @@ class KotlinRenderer internal constructor(
+                is TableBody -> visit(customNode)
+                is TableRow -> visit(customNode)
+                is TableCell -> visit(customNode)
+                is Strikethrough -> visit(customNode)
```
Nit: Use alphabetical order here

Side note: In general, my rule is fields should be ordered in SOME way. This prevents multiple people on a team always just appended to the end of a list by default, which increases the chance of incurring merge requests especially on larger teams.

Alphabetization is not necessarily the best ordering -- usually, grouping fields by their relationship is more appropriate. However, alphabetization is a decent fallback and is better than nothing.

bitspittle commented on tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/KotlinRenderer.kt:
```diff
@@ -483,6 +473,58 @@ class KotlinRenderer internal constructor(
+                    val scope = NodeScope(reporter, data, indentCount)
+                    val label = customNode.label
+                    // Create a link to the footnote definition
+                    val code = "$KOBWEB_DOM.GenericTag(\"sup\", \"id=\\\"fnref-$label\\\"\")"
```
Handler code has ended up inside renderer code and that doesn't seem correct. KotlinRenderer should mostly be concerned with visiting Markdown nodes and delegating logic to the handlers. 

For example, if someone wanted to override the look and feel of footnotes in their own site, I don't think they can do it anymore.

bitspittle commented on tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/KotlinRenderer.kt:
```diff
@@ -483,6 +473,58 @@ class KotlinRenderer internal constructor(
+                    // Render a task list item, which is a checkbox that is either checked or not.
+                    // We prefer to use Silk's FontAwesome icons if available, otherwise we fall back to a
+                    // standard HTML checkbox.
+                    val code = if (handlers.useSilk.get()) {
```
Yeah, KotlinRenderer should know nothing about silk. (Imagine a user in the future uses some third party UI library; there's no way we could handle all future possible cases)

bitspittle commented on playground/site/src/jsMain/resources/markdown/documents/unified_test.md:
```diff
@@ -0,0 +1,102 @@
+| Task Lists       | ✅      | Creates checklists |
+| HTML in Markdown | ✅      | Allows embedding raw HTML tags |
+
+
```
Optional: Remove excessive newlines?

bitspittle reviewed (commented):
I really appreciate you getting your hands dirty and jumping into some of the uglier code in the Kobweb project :)

Taking a look at your PR, I think the biggest issue overall is some handler code ending up inside renderer code, and some parsing logic ending up inside handlers.

It should go: parser (focused only on markdown) -> renderer (focused mostly on visiting nodes) -> handlers (focused on providing Kotlin/JS code), something like that. (It's been a while, my memory is a bit hazy, but if you see some code that seems to contradict this, feel free to call attention to it)

I haven't looked too deeply at the proposed code yet to see if I could suggest a better way. I think for now I'll just leave comments mentioning my concerns and, if you can't see a way to resolve them (say, my bad code won't allow it), let me know and we can go back and forth.

For now, I would recommend simplifying this PR to *just* strikethrough, because that seems like a relatively simple extension. And we can tackle the footnote stuff / comprehensive tests later.

Also, I'll do the next release without these markdown changes since they're a bit more involved than I thought; we can get them ready for the release after that. That way, there's no need to feel any pressure to rush.

kocheick committed:
update: Adjust unified_test.md routeOverride, location and remove unused layout property

kocheick committed:
add support for footnotes and enhance task list rendering in markdown

kocheick committed:
enable footnotes support and improve task list rendering in markdown

kocheick committed:
Update code changes

kocheick committed:
bumped commonmark from 0.24.0 to 0.25.1

kocheick committed:
location update

kocheick committed:
location update

kocheick committed:
refactor: Reorganize MarkdownFeatures properties and extension setup

- Consolidate `footnotes` and `strikethrough` properties for clarity
- Adjust initialization sequence for extensions to ensure proper loading and maintain code readability

kocheick committed:
Merge branch 'dev' into feature/enhanced-markdown-support

# Conflicts:
#	playground/site/src/jsMain/resources/markdown/Markdown.md
#	tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/KotlinRenderer.kt
#	tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/MarkdownFeatures.kt
#	tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/handlers/MarkdownHandlers.kt

bitspittle commented on tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/handlers/MarkdownHandlers.kt:
```diff
@@ -436,6 +489,30 @@ abstract class MarkdownHandlers @Inject constructor(project: Project) {
+
+        footnoteReference.convention { reference ->
+            val label = reference.label
+            "$KOBWEB_DOM.GenericTag(\"sup\", \"id=\\\"fnref-$label\\\"\") { $KOBWEB_DOM.GenericTag(\"a\", \"href=\\\"#fn-$label\\\"\") { $JB_DOM.Text(\"$label\") } }"
```
If we're going to add an id, maybe "kobweb-footnote-$label"? 

Have you tested what happens if you have multiple duplicate references with the same label?

kocheick replied:
multiple duplicate can reference the same label. we do have a test example at `playground/site/src/jsMain/resources/markdown/tests/unified_test.md`

bitspittle reviewed (changes requested):
I haven't run your code yet, I will after the next round of comments. But this definitely looks way cleaner than the first approach, thanks for taking the time to refactor.

kocheick committed:
Merge branch 'dev' into feature/enhanced-markdown-support

kocheick committed:
enable footnotes and improve markdown rendering in unified_test.md

kocheick committed:
add support for footnotes and enhance task list rendering in markdown

kocheick committed:
remove deprecated markdown test document

kocheick committed:
Merge remote-tracking branch 'origin/feature/enhanced-markdown-support' into feature/enhanced-markdown-support

kocheick committed:
fix yaml quote escaping to prevent breaking generated code

kocheick commented on tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/handlers/MarkdownHandlers.kt:
```diff
@@ -436,6 +489,30 @@ abstract class MarkdownHandlers @Inject constructor(project: Project) {
+            }
+        }
+
+        footnoteDefinition.convention { definition ->
```
ill reorganize the handlers node in the init block

kocheick committed:
fix yaml quote escaping to prevent broken generated code

kocheick commented on tools/gradle-plugins/extensions/markdown/src/main/kotlin/com/varabyte/kobwebx/gradle/markdown/handlers/MarkdownHandlers.kt:
```diff
@@ -436,6 +489,30 @@ abstract class MarkdownHandlers @Inject constructor(project: Project) {
+            val isChecked = marker.isChecked
+            if (useSilk.get()) {
+                val iconCode = if (isChecked) {
+                    "com.varabyte.kobweb.silk.components.icons.fa.FaSquareCheck"
```
when I use **SILK** 
```
       taskListItemMarker.convention { marker ->
           "com.varabyte.kobweb.silk.components.forms.Checkbox(checked = $marker.isChecked, enabled = false, onCheckedChange = {})"
        }
```
the checkbox is as in the below image and its generated content :
```
<li>
   <label class="silk-checkbox kobweb-row kobweb-arrange-start kobweb-align-center-vert silk-disabled silk-checkbox-size_md" aria-disabled="true" tabindex="-1">
      <input class="silk-input silk-input-checkbox silk-input-size_md" type="checkbox" aria-disabled="true" disabled="" spellcheck="false">
      <div class="silk-checkbox-icon-container silk-checkbox-icon-container-checked kobweb-box kobweb-align-center">
         <div class="silk-checkbox-icon kobweb-box kobweb-align-center">
            <svg style="overflow: visible;" width="1em" viewBox="0 0 24 20" stroke="currentColor" fill="none" stroke-width="4">
               <polyline points="3,12 9,18 21,2"></polyline>
            </svg>
         </div>
      </div>
   </label>
   Completed task
</li>
```

<img width="411" height="310" alt="SILK Screensho" src="https://github.com/user-attachments/assets/ff569770-d857-44fa-a492-0fbe86746689" />

---

While when I do use **KOBWEB_DOM**
```
       taskListItemMarker.convention { marker ->
                        "$KOBWEB_DOM.GenericTag(\"input\", \"type=\\\"checkbox\\\" ${if (isChecked) "checked " else ""}disabled\")"
        }
```
 it gets rendered correctly, see image below and its generated content  :
`<li><input type="checkbox" checked="" disabled="">Completed task</li>`

<img width="411" height="251" alt="KOBWEB_DOM Screenshot" src="https://github.com/user-attachments/assets/26690989-37bf-446f-97f1-285ff0e86d96" />

 

I'm not sure why SILK isn't displaying things right, or if this was intended. Let me know if you would still recommend SILK only or  maybe KOBWEB_DOM (Compose).

 I initially thought doing this would be safe but I don't have a strong rational for it 
```        
taskListItemMarker.convention { marker ->
            val isChecked = marker.isChecked
            if (useSilk.get()) 
                "$SILK.forms.Checkbox(checked = $isChecked, enabled = false, onCheckedChange = {})"
            else
                "$KOBWEB_DOM.GenericTag(\"input\", \"type=\\\"checkbox\\\" ${if (isChecked) "checked " else ""}disabled\")"
        }
```

kocheick committed:
Replaced GenericTag with Compose HTML components wherever possible

kocheick committed:
refactor markdown handler to conditionally add classes for task list items and containers
